src/formattedcode/output_cyclonedx.py

```diff
@@ -0,0 +1,659 @@
+                                   help_group=OUTPUT_GROUP,
+                                   sort_order=70),
+    ]
+
+    def is_enabled(self, output_cyclonedx, **kwargs):
+        return output_cyclonedx
+
+    def process_codebase(self, codebase, output_cyclonedx, **kwargs):
+        bom = build_bom(codebase)
+        write_results(bom, output_cyclonedx, output_json=False)
+
+
+@output_impl
+class CycloneDxJsonOutput(OutputPlugin):
+    options = [
+        PluggableCommandLineOption(('--cyclonedx-json', 'output_cyclonedx_json',),
+                                   type=FileOptionType(
+                                       mode='w', encoding='utf-8', lazy=True),
+                                   metavar='FILE',
+                                   help='Write scan output in CycloneDX JSON format to FILE.',
+                                   help_group=OUTPUT_GROUP,
+                                   sort_order=70),
+    ]
+
+    def is_enabled(self, output_cyclonedx_json, **kwargs):
+        return output_cyclonedx_json
+
+    def process_codebase(self, codebase, output_cyclonedx_json, **kwargs):
+        bom = build_bom(codebase)
+        write_results(bom, output_file=output_cyclonedx_json)
+
+
+def _get_dependency_candidate(dependency, candidates):
+    if len(candidates) == 1:
+        return list(candidates.values())[0]
+    else:
+        return None
+
+
+def _set_key_or_append(dictionary: dict, key, value):
+    if key in dictionary:
+        dictionary[key].append(value)
+    else:
+        dictionary[key] = [value]
+
+
+def generate_dependencies_list(dep_map, comp_map) -> List[CycloneDxDependency]:
+    # return early if we have no components to operate on
+    if len(comp_map) == 0:
+        return None
+
+    # holds a mapping of type purl -> list(purl)
+    dependencies = {}
+
+    resolved_scopes = {}
+
+    for purl in dep_map:
+        for dependency in dep_map[purl]:
+            if dependency.get("is_resolved"):
+                _set_key_or_append(dependencies, purl, dependency.get("purl"))
+            else:
+                candidates = dict(filter(
+                    lambda item: item[0] and item[0].startswith(dependency["purl"]),
+                    comp_map.items()))
+                resolved_dependency = _get_dependency_candidate(
+                    dependency, candidates)
+                if resolved_dependency is not None:
+                    _set_key_or_append(dependencies, purl, resolved_dependency.bom_ref)
+                    _set_key_or_append(resolved_scopes, purl, dependency.get("is_optional"))
+
+    for purl in resolved_scopes:
+        if all(resolved_scopes[purl]):
+            comp_map[purl].scope=CycloneDxComponentScope.OPTIONAL
+
+
+    return list(
+        map(
+            lambda entry:
+            CycloneDxDependency(ref=entry[0], dependsOn=entry[1]),
+            dependencies.items()
+        )
+    )
+
+
+def merge_components(existing: CycloneDxComponent, new: CycloneDxComponent):
```

Comment:
Since we use the `purl` as a `bom-ref` it has to be unique, merging boms that have the same purl ensures this.  
This method now checks if the `purl` values are indeed identical before merging and raises a `ValueError` if they are not.

Please note that the only location where we call this helper doesn't catch that because the context in which it is invoked already ensures that both components have the same purl

Reply:
This may be a problem. ScanCode reports all instances of each package so there can be duplicated purls (and this is evolving further with https://github.com/nexB/scancode-toolkit/pull/2748 )
The thing is that each package instance may have the same purl but can different attributes (for instanced when derived from a certain manifest vs. another format of style such as a Gemfile, gemspec and installed metadata.